pkg/controller/nodes/predicate.go

```diff
@@ -62,21 +62,21 @@ func CanExecute(ctx context.Context, dag executors.DAGStructure, nl executors.No
 			return PredicatePhaseNotReady, nil
 		}
 
-		if parentNodeID != nil && *parentNodeID == upstreamNodeID {
-			upstreamNode, ok := nl.GetNode(upstreamNodeID)
-			if !ok {
-				return PredicatePhaseUndefined, errors.Errorf(errors.BadSpecificationError, nodeID, "Upstream node [%v] of node [%v] not defined", upstreamNodeID, nodeID)
-			}
+		upstreamNode, ok := nl.GetNode(upstreamNodeID)
+		if !ok {
+			return PredicatePhaseUndefined, errors.Errorf(errors.BadSpecificationError, nodeID, "Upstream node [%v] of node [%v] not defined", upstreamNodeID, nodeID)
+		}
 
-			// Deprecated: This if block will be removed in a future version. It's harmless (will be no-op) for newly
-			// compiled Workflows as sub-branch-nodes won't have an execution or code dependency on branch nodes.
-			// This only happens if current node is the child node of a branch node
-			if upstreamNode.GetBranchNode() == nil || upstreamNodeStatus.GetBranchStatus().GetPhase() != v1alpha1.BranchNodeSuccess {
-				logger.Debugf(ctx, "Branch sub node is expected to have parent branch node in succeeded state")
-				return PredicatePhaseUndefined, errors.Errorf(errors.IllegalStateError, nodeID, "Upstream node [%v] is set as parent, but is not a branch node of [%v] or in illegal state.", upstreamNodeID, nodeID)
+		if upstreamNode.GetBranchNode() != nil && upstreamNodeStatus.GetBranchStatus() != nil {
```

Comment:
I think the condition is still puzzling... I understand what it's trying to do but I think we should rephrase it like below. The issue I take is that we check if upstream is a branch node... which technically we don't care about UNLESS it's also the parent node... I understand the condition will work either way but it's just odd to special case branch node in that case... does that makes sense?

```
If I've a parentId and it's the same as the upstreamNode:
   If it's a branch node, and status is not BranchNodeSuccess, return not ready
   If it's a branch node, and status is BranchNodeSuccess, continue
   If it's NOT a branch node, error? should never happen? 
Else: check upstream has succeeded... etc. etc.
```